datafusion/functions/src/string/concat.rs

```diff
@@ -65,14 +75,223 @@ impl Default for ConcatFunc {
 
 impl ConcatFunc {
     pub fn new() -> Self {
-        use DataType::*;
         Self {
-            signature: Signature::variadic(
-                vec![Utf8View, Utf8, LargeUtf8],
-                Volatility::Immutable,
-            ),
+            signature: Signature::user_defined(Volatility::Immutable),
         }
     }
+
+    fn concat_arrays(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+        if args.is_empty() {
+            return plan_err!("concat requires at least one argument");
+        }
+
+        // Simple case: single row - use fast path
+        let num_rows = args
+            .iter()
+            .find_map(|arg| match arg {
+                ColumnarValue::Array(array) => Some(array.len()),
+                _ => None,
+            })
+            .unwrap_or(1);
```

Comment:
I feel this count could be obtained from the original ScalarFunctionArgs and passed through, instead of having this logic (which doesn't account for scalars)